lib/honeybadger/event_context.ex

```diff
@@ -0,0 +1,83 @@
+defmodule Honeybadger.EventContext do
+  @moduledoc false
+
+  @key __MODULE__
+
+  @doc """
+  Merge the given map or keyword list into the current event context.
+  """
+  @spec merge(map() | list()) :: map()
+  def merge(kw) when is_list(kw), do: merge(Map.new(kw))
+
+  def merge(context) when is_map(context) do
+    new_context = Map.merge(get(), context)
+    Process.put(@key, new_context)
+
+    new_context
+  end
+
+  @doc """
+  Replace the current event context with the given map or keyword list.
+  """
+  @spec replace(map() | list()) :: map()
+  def replace(kw) when is_list(kw), do: replace(Map.new(kw))
+
+  def replace(context) when is_map(context) do
+    Process.put(@key, context)
+
+    context
+  end
+
+  @doc """
+  Put a new key-value pair in the current event context if the key does not
+  already exist. You can lazy initialize the value by passing a function that
+  returns the value. The function will only be called if the key does not exist
+  in the current event context.
+  """
+  @spec put_new(atom(), (-> any()) | any()) :: map()
+  def put_new(key, f) when is_function(f, 0) do
+    new_context = Map.put_new_lazy(get(), key, f)
```

Comment:
Nitpick only, to avoid single letter variable:

```suggestion
  def put_new(key, fun) when is_function(fun, 0) do
    new_context = Map.put_new_lazy(get(), key, fun)
```